src/Chainweb/Pact/Backend/DbCache.hs

```diff
@@ -0,0 +1,236 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Module: Chainweb.Pact.Backend.DbCache
+-- Copyright: Copyright © 2022 Kadena LLC.
+-- License: See LICENSE file
+--
+-- LRU cache for avoiding expensive JSON deserialization.
+--
+module Chainweb.Pact.Backend.DbCache
+( DbCache
+, checkDbCache
+, emptyDbCache
+, cacheSize
+, cacheCount
+, isEmptyCache
+, cacheStats
+, updateCacheStats
+) where
+
+import Control.Lens hiding ((.=))
+import Control.Monad.State.Strict
+
+import qualified Crypto.Hash as C (hash)
+import Crypto.Hash.Algorithms
+
+import Data.Aeson (FromJSON, Value, object, decodeStrict, (.=))
+import qualified Data.ByteArray as BA
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Short as BS
+import Data.Hashable
+import qualified Data.HashMap.Strict as HM
+import Data.List (sort)
+
+import Database.SQLite3.Direct
+
+import GHC.Compact
+
+import Pact.Types.Persistence
+
+-- -------------------------------------------------------------------------- --
+-- CacheAddress
+
+newtype CacheAddress = CacheAddress BS.ShortByteString
+    deriving (Show,Eq,Ord,Hashable)
+
+-- -------------------------------------------------------------------------- --
+-- CacheEntry
+
+data CacheEntry a = CacheEntry
+  { _ceTxId :: !TxId
+    -- ^ Priority for cache eviction
+  , _ceAddy :: !CacheAddress
+    -- ^ Key. Uniquely identifies entries in the cache
+  , _ceSize :: !Int
+    -- ^ The size of the entry. Used to keep track of the cache limit
+  , _ceData :: !a
+    -- ^ Cached data. This is stored in a compact region.
+  , _ceHits :: Int
+    -- ^ Count of cache hits for this entry
+  }
+
+ceTxId :: Lens' (CacheEntry a) TxId
+ceTxId = lens _ceTxId (\a b -> a { _ceTxId = b })
```

Comment:
Is there a reason we don't want to use `makeLenses ''CacheEntry` here?

Reply:
We would have to export it from the module (which we don't want) or GHC would warn about unused symbols. I guess there's a way to only generate some lenses. But, I don't think it's worth adding TH overhead for those four lines of code.